include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -840,6 +840,171 @@ class __static_monotonic_dispatcher<::std::integer_sequence<::std::uint16_t, _X,
+    std::size_t __full_range_size;
+
+    template <typename _IdxType, typename _WalkFunction, typename... _Rngs>
+    void
+    operator()(std::true_type, _IdxType __idx, _WalkFunction __f, _Rngs&&... __rngs) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+        {
+            __f(__rngs[__idx + __i]...);
+        }
+    }
+    // For a non-full vector path, process it sequentially. This will always be the last sub or work group
+    // if it does not evenly divide into input
+    template <typename _IdxType, typename _WalkFunction, typename... _Rngs>
+    void
+    operator()(std::false_type, _IdxType __idx, _WalkFunction __f, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+        {
+            __f(__rngs[__idx + __i]...);
+        }
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_store
+{
+    static_assert(__vec_size <= 4, "Only vector sizes of 4 or less are supported");
+    std::size_t __full_range_size;
+
+    template <typename _IdxType, typename _StoreOp, typename... _Rngs>
+    void
+    operator()(std::true_type, _IdxType __start_idx, _StoreOp __store_op, _Rngs&&... __rngs) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+            __store_op(__i, __start_idx + __i, __rngs...);
+    }
+    template <typename _IdxType, typename _StoreOp, typename... _Rngs>
+    void
+    operator()(std::false_type, _IdxType __start_idx, _StoreOp __store_op, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __start_idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+            __store_op(__i, __start_idx + __i, __rngs...);
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_reverse
+{
+    static_assert(__vec_size <= 4, "Only vector sizes of 4 or less are supported");
+    template <typename _Idx, typename _Array>
+    void
+    operator()(/*__is_full*/ std::true_type, const _Idx /*__elements_to_process*/, _Array __array) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size / 2; ++__i)
+            std::swap(__array[__i], __array[__vec_size - __i - 1]);
+    }
+    template <typename _Idx, typename _Array>
+    void
+    operator()(/*__is_full*/ std::false_type, const _Idx __elements_to_process, _Array __array) const
+    {
+        for (std::uint8_t __i = 0; __i < __elements_to_process / 2; ++__i)
+            std::swap(__array[__i], __array[__elements_to_process - __i - 1]);
+    }
+};
+
+// Processes a loop with a given stride. Intended to be used with sub-group / work-group strides for good memory access patterns
+// (potentially with vectorization)
+template <std::uint8_t __num_strides>
```

Comment:
Even `__strided_loop` could probably be grouped with the others I mention above `__vector_[load,store,walk] with the appropriate abstraction.  Either a multiply in the functor or a stride argument to the baseline shared helper.

Reply:
Added to the TODO comment.